### DISCUSSION TITLE
Is there any docs to introduce how the ‘Named configurations’ works ?

### DISCUSSION THREAD

**Discussion by ZCShou** [ipxe/ipxe#703] [Q&A]
the doc of  'https://ipxe.org/appnote/named_config' is tool simple.

In other words, where is the ipxe's detail docs about architecture, config and so on ? The doc(based doxygen) in the source is not useful at all.

**Comment by NiKiZe** [comment]
I think https://ipxe.org/appnote/named_config explains exactly how it works. Could you elaborate on what you feel is missing?

Have you also looked in the existing well documented config/*.h files?
Maybe there is an incorrect assumption on how things work? But please help us better understand the request.

**Reply by ZCShou** [reply]
Thanks for your quickly reply.

When building with `CONFIG=<name>`  ,  should I build `<name>` firstly? And what is difference between `CONFIG` and `LOCAL_CONFIG` in the `named.h`?

The iPXE.org says "It is possible to build binaries with different configurations from the same codebase using [named configurations](https://ipxe.org/appnote/named_config). This makes it easy to have one configuration for a ROM, another one for a chainloaded binary and a third configuration for a UEFI binary.", if I want to build three above, how should I do.

**Reply by ZCShou** [reply]
I think:
`config/<name>` ,likely `vbox`, is your predefined for some case, but users may not use them.
`config/local/<name>` is built by myself, in where I can use to store my config file.
Is right?

**Reply by ZCShou** [reply]
Maybe like this.
![iPXE-CONFIG](https://user-images.githubusercontent.com/8591065/184085169-ef8d3877-fdac-44ec-aa94-0f5d28f759e9.png)

**Reply by NiKiZe** [reply]
Buildtargets eg. bin/undionly.kpxe is not connected to named configs.

make bin/ipxe.rom CONFIG=vbox would make sense. But nothing is predefined in terms of coronation between buildtarget and named config.

**Comment by thesefer** [comment]
Hi, wanting to do it properly I tried to start using a named configuration for my target. I could have sworn that yesterday it was working but today trying to rebuild it I am in an endless `[DEPS]` loop. Even if I leave all files empty it loops itself. I think I am doing it wrong?

Command i.e.: `make bin-x86_64-efi/ipxe.efi CONFIG=uefi-test`

I created the folder `config/uefi-test` and added two files there which I like to modify. :
console.h
```
#define	CONSOLE_FRAMEBUFFER	/* Graphical framebuffer console */
#undef KEYBOARD_MAP
#define	KEYBOARD_MAP de
```

general.h
```
#define	DOWNLOAD_PROTO_HTTPS	/* Secure Hypertext Transfer Protocol */
#define	IMAGE_SCRIPT		/* iPXE script image support */
#define	IMAGE_EFI		/* EFI image support */
#define	IMAGE_PNG		/* PNG image support */
#define	IMAGE_DER		/* DER image support */
#define	IMAGE_PEM		/* PEM image support */
#define NSLOOKUP_CMD		/* DNS resolving command */
#define TIME_CMD		/* Time commands */
#define REBOOT_CMD		/* Reboot command */
#define POWEROFF_CMD		/* Power off command */
#define PARAM_CMD		/* Request parameter commands */
#define PING_CMD		/* Ping command */
#define CONSOLE_CMD		/* Console command */
#define NTP_CMD		/* NTP commands */
```

**Comment by mcb30** [comment]
On 29/10/2024 11:00, thesefer wrote:
> Hi, wanting to do it properly I tried to start using a named 
> configuration for my target. I could have sworn that yesterday it was 
> working but today trying to rebuild it I am in an endless |[DEPS]| loop. 
> Even if I leave all files empty it loops itself. I think I am doing it 
> wrong?
> 
> Command i.e.: |make bin-x86_64-efi/ipxe.efi CONFIG=uefi-test|
> 
> I created the folder |config/uefi-test| and added two files there which 
> I like to modify. :
> console.h
> general.h

You'll also need to create .h files for every other header that can be 
part of a named config (even if the .h file is empty).  You can see 
which ones are needed by looking in an existing named config directory, 
e.g. config/qemu.  The current list is:

   colour.h
   console.h
   crypto.h
   general.h
   ioapi.h
   serial.h
   settings.h
   sideband.h
   usb.h

The documentation on this is kind of sparse, sorry.  Happy to accept 
edits to https://ipxe.org/appnote/named_config (which should be editable 
to anyone with a user account, although there may be spam moderation 
issues to deal with).

Thanks,

Michael

**Reply by thesefer** [reply]
Thank you very much. I must have cleaned everything in local, now that I remember from my transition to named configuration!
